### PR TITLE
Refactor TicketSerializer/PropertiesSerializer and add ClaimsIdentity.Actor/Claim.Properties support

### PR DESCRIPTION
**Pull request by PinpointTownes**
Fixes https://github.com/aspnet/Security/issues/324, partially fixes https://github.com/aspnet/Security/issues/323 and deprecates https://github.com/aspnet/Security/issues/158.

@Eilon any chance we could include this PR in beta8 release?

/cc @HaoK @Tratcher

### CODE REVIEW

File: src/Microsoft.AspNet.Authentication/DataHandler/PropertiesSerializer.cs [hunk 1]
```diff
@@ -39,26 +41,29 @@ public AuthenticationProperties Deserialize(byte[] data)
...
         {
             writer.Write(FormatVersion);
             writer.Write(properties.Items.Count);
-            foreach (var kv in properties.Items)
+
+            foreach (var item in properties.Items)
             {
-                writer.Write(kv.Key);
-                writer.Write(kv.Value);
+                writer.Write(item.Key ?? string.Empty);
```
Comment (PinpointTownes) [inline src/Microsoft.AspNet.Authentication/DataHandler/PropertiesSerializer.cs]: Fixes the bug seen in https://github.com/aspnet-contrib/AspNet.Security.OpenIdConnect.Server/issues/126.

**Comment by PinpointTownes** [comment]
@brentschmaltz question: should we also support `ClaimsIdentity.Label` serialization/deserialization?

File: src/Microsoft.AspNet.Authentication/DataHandler/TicketSerializer.cs [hunk 2]
```diff
@@ -36,98 +39,177 @@ public virtual AuthenticationTicket Deserialize(byte[] data)
...
+                identity.BootstrapContext = reader.ReadString();
             }
 
-            var properties = PropertiesSerializer.Read(reader);
-            return new AuthenticationTicket(new ClaimsPrincipal(identities), properties, authenticationScheme);
+            // Determine whether the identity
+            // has an actor identity attached.
+            if (reader.ReadBoolean())
+            {
+                identity.Actor = ReadIdentity(reader);
```
Comment (PinpointTownes) [inline src/Microsoft.AspNet.Authentication/DataHandler/TicketSerializer.cs]: Note: `ClaimsIdentity.Actor` is guarded against circular references, so there's no risk to end up in an endless recursive call :smile:

File: src/Microsoft.AspNet.Authentication/DataHandler/TicketSerializer.cs [hunk 3]
```diff
@@ -36,98 +39,177 @@ public virtual AuthenticationTicket Deserialize(byte[] data)
...
+        public virtual void WriteClaim([NotNull] BinaryWriter writer, [NotNull] Claim claim)
+        {
+            WriteWithDefault(writer, claim.Type, claim.Subject?.NameClaimType ?? ClaimsIdentity.DefaultNameClaimType);
+            writer.Write(claim.Value);
+            WriteWithDefault(writer, claim.ValueType, ClaimValueTypes.String);
+            WriteWithDefault(writer, claim.Issuer, ClaimsIdentity.DefaultIssuer);
+            WriteWithDefault(writer, claim.OriginalIssuer, claim.Issuer);
+
+            // Write the number of properties contained in the claim.
+            writer.Write(claim.Properties.Count);
```
Comment (PinpointTownes) [inline src/Microsoft.AspNet.Authentication/DataHandler/TicketSerializer.cs]: `Claim.Properties` is not supposed to be null, but old versions of <Mono 4.0 were known to not initialize this property. I chatted with @davidfowl and he agreed it was not worth it.

**Comment by Eilon** [comment]
@Tratcher thoughts about taking this?

File: src/Microsoft.AspNet.Authentication/DataHandler/DataSerializers.cs [hunk 4]
```diff
@@ -9,8 +9,8 @@ public static class DataSerializers
     {
         static DataSerializers()
         {
-            Properties = new PropertiesSerializer();
-            Ticket = new TicketSerializer();
+            Properties = PropertiesSerializer.Default;
```
Comment (HaoK) [inline src/Microsoft.AspNet.Authentication/DataHandler/DataSerializers.cs]: I think I nuked these statics in commit today

Comment (PinpointTownes) [inline src/Microsoft.AspNet.Authentication/DataHandler/DataSerializers.cs]: Yes, you ruined my PR :cry: 

Bad Hao, bad, bad, bad :smile:

**Comment by PinpointTownes** [comment]
Rebased, but the build failed because of this super annoying `Invalid chunked data` error (that has nothing to do with this PR). I'll restart it tomorrow.

**Comment by PinpointTownes** [comment]
@brentschmaltz last call: do you think we should also serialize ClaimsIdentity.Label?

File: src/Microsoft.AspNet.Authentication/DataHandler/TicketSerializer.cs [hunk 5]
```diff
@@ -36,98 +39,177 @@ public virtual AuthenticationTicket Deserialize(byte[] data)
             }
         }
 
-        public static void Write([NotNull] BinaryWriter writer, [NotNull] AuthenticationTicket model)
+        public virtual void Write([NotNull] BinaryWriter writer, [NotNull] AuthenticationTicket ticket)
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication/DataHandler/TicketSerializer.cs]: Why are any of these read/write methods public? At best they'd be protected.

Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication/DataHandler/TicketSerializer.cs]: If it's just for tests then they should be protected internal.

Comment (PinpointTownes) [inline src/Microsoft.AspNet.Authentication/DataHandler/TicketSerializer.cs]: I haven't changed the visibility modifier of these methods, but this is something we can consider.

That said, note that `PropertiesSerializer.Write/Read` is internally used by the Twitter middleware.
If you want to make everything protected internal, we'll have to add a `InternalsVisibleToAttribute` for both the test project and the Twitter provider.

Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication/DataHandler/TicketSerializer.cs]: Ah, I missed the Twitter cross-over. That's pretty sloppy, but it doesn't look easy to fix. Disregard.

Comment (PinpointTownes) [inline src/Microsoft.AspNet.Authentication/DataHandler/TicketSerializer.cs]: We can keep Write/Read public and make WriteIdentity/WriteClaim/ReadIdentity/ReadClaim protected if you want.

Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication/DataHandler/TicketSerializer.cs]: That's better than nothing.

Comment (PinpointTownes) [inline src/Microsoft.AspNet.Authentication/DataHandler/TicketSerializer.cs]: Done.

**Comment by Tratcher** [comment]
:shipit:

**Comment by Tratcher** [comment]
Rebased and merged.

**Pull request closed without merging**